File: 프로그래머스/3/72414. 광고 삽입/광고 삽입.py
def str2int(time):
    h = int(time[:2])
    m = int(time[3:5])
    s = int(time[6:])
    
    return h * 3600 + m * 60 + s

def int2str(time):
    h = time // 3600
    time %= 3600
    m = time // 60
    time %= 60
    s = time
    
    return str(h).zfill(2) + ":" + str(m).zfill(2) + ":" + str(s).zfill(2)

def solution(play_time, adv_time, logs):
    play = str2int(play_time)
    prefix_sum = [0] * (play + 1)
    for log in logs:
        start = str2int(log[:8])
        end = str2int(log[9:])
        
        prefix_sum[start] += 1
        prefix_sum[end] -= 1
    
    for i in range(play + 1): prefix_sum[i] += prefix_sum[i - 1] # 초마다 시청자 수 계산
    for i in range(play + 1): prefix_sum[i] += prefix_sum[i - 1] # 누적합으로 변경
    
    adv = str2int(adv_time)
    max_start_time = 0
    max_accm = prefix_sum[adv - 1]
    for i in range(1, play - adv + 1):
        if max_accm < prefix_sum[i + adv - 1] - prefix_sum[i - 1]:
            max_accm = prefix_sum[i + adv - 1] - prefix_sum[i - 1]
            max_start_time = i
    
    return int2str(max_start_time)

File: 프로그래머스/3/72414. 광고 삽입/광고 삽입.cpp
#include <string>
#include <vector>

using namespace std;

int str2int(string time){
    int h = stoi(time.substr(0, 2));
    int m = stoi(time.substr(3, 2));
    int s = stoi(time.substr(6, 2));
    
    return h * 3600 + m * 60 + s;
}

string int2str(int time){
    int h = time / 3600;
    time %= 3600;
    int m = time / 60;
    time %= 60;
    int s = time;
    
    return string(2 - to_string(h).size(), '0') + to_string(h) + ":" +
           string(2 - to_string(m).size(), '0') + to_string(m) + ":" +
           string(2 - to_string(s).size(), '0') + to_string(s);
}

string solution(string play_time, string adv_time, vector<string> logs) {
    int play = str2int(play_time);
    vector<long long> prefix_sum(play + 1, 0);
    for(string log : logs){
        int start = str2int(log.substr(0, 8));
        int end = str2int(log.substr(9));
        
        prefix_sum[start]++;
        prefix_sum[end]--;
    }
    
    for(int i = 1; i <= play; i++) prefix_sum[i] += prefix_sum[i - 1]; // 초마다 시청자 수 계산
    for(int i = 1; i <= play; i++) prefix_sum[i] += prefix_sum[i - 1]; // 누적합으로 만듦
    
    int adv = str2int(adv_time);
    int max_start_time = 0;
    long long max_accm = prefix_sum[adv - 1];
    for(int i = 1; i + adv <= play; i++){
        if(max_accm < prefix_sum[i + adv - 1] - prefix_sum[i - 1]){
            max_accm = prefix_sum[i + adv - 1] - prefix_sum[i - 1];
            max_start_time = i;
        }
    }
    return int2str(max_start_time);
}